crates/namada/src/ledger/native_vp/masp.rs

```diff
@@ -279,66 +299,385 @@ where
+        );
+        // Public keys must be the hash of the sources/targets
+        let address_hash = addr_taddr(counterpart.clone());
+        // Enable the decoding of these counterpart addresses
+        result
+            .decoder
+            .insert(address_hash, TAddrData::Addr(counterpart.clone()));
+        // Finally record the actual balance change starting with the initial
+        // state
+        let pre_entry =
+            result.pre.entry(address_hash).or_insert(ValueSum::zero());
+        *pre_entry = checked!(
+            pre_entry + &ValueSum::from_pair((*token).clone(), pre_balance)
+        )
+        .map_err(native_vp::Error::new)?;
+        // And then record thee final state
+        let post_entry =
+            result.post.entry(address_hash).or_insert(ValueSum::zero());
+        *post_entry = checked!(
+            post_entry + &ValueSum::from_pair((*token).clone(), post_balance)
+        )
+        .map_err(native_vp::Error::new)?;
+        Result::<_>::Ok(result)
+    }
+
+    // Collect the various packet acknowledgements and store them by their
+    // commitments
+    fn extract_ibc_acknowledgements(
+        &self,
+    ) -> Result<BTreeMap<Vec<u8>, AcknowledgementStatus>> {
+        let mut ibc_acks = BTreeMap::new();
+        let ibc_events = self.ctx.state.write_log().get_events_of::<IbcEvent>();
+        for ibc_event in ibc_events {
+            if ibc_event.kind().sub_domain() == WRITE_ACK_EVENT {
+                // Packet acknowledgements are an attribute of the IBC Write
+                // Acknowledgement event
+                let acknowledgement = ibc_event
+                    .raw_read_attribute::<PacketAck<'_>>()
+                    .ok_or_err_msg(
+                        "packet_ack attribute missing from \
+                         write_acknowledgement",
+                    )?;
+                // Acknowledgements adhere to the AcknowledgementStatus format
+                let acknowledgement: AcknowledgementStatus =
+                    serde_json::from_slice(acknowledgement.as_ref())
+                        .wrap_err("Decoding acknowledment failed")?;
+                let commitment =
+                    compute_ack_commitment(&acknowledgement.clone().into());
+                // Store the acknowledgements by their commitment to facilitate
+                // reverse lookups on the commitments placed in storage
+                ibc_acks.insert(commitment.into_vec(), acknowledgement);
+            }
+        }
+        Ok(ibc_acks)
+    }
+
     // Check that transfer is pinned correctly and record the balance changes
     fn validate_state_and_get_transfer_data(
         &self,
         keys_changed: &BTreeSet<Key>,
+        ibc_msgs: &[IbcMessage],
     ) -> Result<ChangedBalances> {
         // Get the changed balance keys
-        let mut counterparts_balances = keys_changed
-            .iter()
-            .filter_map(is_any_shielded_action_balance_key);
```

Comment:
Can we remove the `is_any_shielded_action_balancekey` if we don't need it anymore?

Reply:
Removed the function and `ShieldedActionOwner`. Thanks.